### PR TITLE
Add PNPM support

### PR DESCRIPTION
**Pull request by deivid-rodriguez**
Just sharing some initial work!

### CODE REVIEW

File: npm_and_yarn/Dockerfile [hunk 1]
```diff
@@ -16,6 +16,10 @@ RUN curl -sL https://deb.nodesource.com/setup_16.x | bash - \
 RUN corepack enable \
   && corepack prepare yarn@$YARN_VERSION --activate
 
+# Install pnpm and set it to a stable version
+RUN corepack enable \
+  && corepack prepare pnpm@8.2.0 --activate
```
Comment (raulfdm) [inline npm_and_yarn/Dockerfile:21]: I know this PR is in the early stage, but just to mention, `pnpm` major versions are most often not backward compatible.

If you have a project running pnpm v7 and tries to install deps with v8, it won't work. They use the `pnpm-lock.yaml` file to control this.

Comment (deivid-rodriguez) [reply npm_and_yarn/Dockerfile:21]: Thanks! Yes, we'll carefully consider the different lockfile formats and PNPM versions. Of course, we'll limit support to some reasonable range of lockfile formats and PNPM versions.

In any case, I'd recommend using the [packageManager](https://nodejs.org/api/packages.html#determining-package-manager) field in `package.json` for best support, since that gives Dependabot a hint of what version of PNPM will produce the most compatible result.

File: npm_and_yarn/lib/dependabot/npm_and_yarn/file_fetcher.rb [hunk 2]
```diff
@@ -418,6 +429,24 @@ def fetch_lerna_packages
...
+
+        package_json_files = []
+
+        workspace_paths(parsed_pnpm_workspace_yaml["packages"]).each do |path|
+          file = File.join(path, "package.json")
+
+          begin
+            package_json_files << fetch_file_from_host(file)
+          rescue Dependabot::DependencyFileNotFound
+            nil
```
Comment (jeffwidman) [inline npm_and_yarn/lib/dependabot/npm_and_yarn/file_fetcher.rb:444]: Is it expected this will happen sometimes?

File: npm_and_yarn/lib/dependabot/npm_and_yarn/file_parser/lockfile_parser.rb [hunk 3]
```diff
@@ -67,6 +68,8 @@ def lockfile_for(file)
           @lockfiles ||= {}
           @lockfiles[file.name] ||= if [*package_locks, *shrinkwraps].include?(file)
                                       JsonLock.new(file)
+                                    elsif pnpm_locks.include?(file)
+                                      PnpmLock.new(file)
                                     else
                                       YarnLock.new(file)
```
Comment (jeffwidman) [inline npm_and_yarn/lib/dependabot/npm_and_yarn/file_parser/lockfile_parser.rb:74]: What about when these cases conflict such that a user has a lock file for both yarn and pnpm... we don't actually know what they prefer us to do, so should we throw an error in that case?

This is probably true other places in this code as well, and while the issue existed previously, it'll be even more true now that there's three options and `pnpm` being such a close cousin of `npm`... if multiple lockfiles are present, we can't definitively say "npm should win, then pnpm, then yarn"...

File: npm_and_yarn/spec/dependabot/npm_and_yarn/file_parser/lockfile_parser_spec.rb [hunk 4]
```diff
@@ -70,6 +70,68 @@
...
+          expect(dependencies.map(&:name)).to_not include("my-fetch-factory")
+        end
+      end
+
+      context "that contain multiple dependencies" do
+        let(:dependency_files) { project_dependency_files("pnpm/no_lockfile_change") }
+
+        its(:length) { is_expected.to eq(370) }
+
+        describe "a repeated dependency" do
```
Comment (jeffwidman) [inline npm_and_yarn/spec/dependabot/npm_and_yarn/file_parser/lockfile_parser_spec.rb:106]: 👍 for this test case

File: npm_and_yarn/spec/dependabot/npm_and_yarn/file_parser/lockfile_parser_spec.rb [hunk 5]
```diff
@@ -70,6 +70,68 @@
...
+          subject { dependencies.find { |d| d.name == "async" } }
+
+          its(:version) { is_expected.to eq("1.5.2") }
+        end
+      end
+
+      context "that contains specified locked to versions with peer disambiguation suffix" do
```
Comment (jeffwidman) [inline npm_and_yarn/spec/dependabot/npm_and_yarn/file_parser/lockfile_parser_spec.rb:113]: typo? reads like there's a missing word, but I'm actually not sure what it might be...

File: npm_and_yarn/spec/dependabot/npm_and_yarn/file_parser/lockfile_parser_spec.rb [hunk 6]
```diff
@@ -70,6 +70,68 @@
...
+          expect(dependencies.map(&:name)).to include("@typescript-eslint/parser")
+        end
+      end
+
+      context "that contain bad lockfile" do
+        let(:dependency_files) { project_dependency_files("pnpm/broken_lockfile") }
+
+        it "raises a DependencyFileNotParseable error" do
+          expect { dependencies }.
+            to raise_error(Dependabot::DependencyFileNotParseable) do |error|
```
Comment (jeffwidman) [inline npm_and_yarn/spec/dependabot/npm_and_yarn/file_parser/lockfile_parser_spec.rb:128]: Oh here you are already expecting the `DependencyFileNotParseable` that I suggested elsewhere... but what's raising it then since the exception is swallowed at that other location?

File: npm_and_yarn/lib/dependabot/npm_and_yarn/file_parser/pnpm_lock.rb [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+            dependency_args = {
+              name: name,
+              version: version,
+              package_manager: "npm_and_yarn",
+              requirements: []
+            }
+
+            if details["dev"]
+              dependency_args[:subdependency_metadata] =
+                [{ production: !details["dev"] }]
```
Comment (jeffwidman) [inline npm_and_yarn/lib/dependabot/npm_and_yarn/file_parser/pnpm_lock.rb:45]: this stanza took a moment to understand what's happening... I wonder if clearer naming of variables might make it more obvious this refers to `dependency-type`'s? `dev`/`production` don't hint at it very strongly... that said, I'm also fine to leave as-is given the helper key name is also `dev`, so viewed through that lens `dev` makes more sense.

File: npm_and_yarn/lib/dependabot/npm_and_yarn/update_checker/version_resolver.rb [hunk 8]
```diff
@@ -332,22 +332,22 @@ def fetch_peer_dependency_errors(version:)
           []
         end
 
-        def handle_peer_dependency_errors(error)
+        def handle_peer_dependency_errors(message)
```
Comment (jeffwidman) [inline npm_and_yarn/lib/dependabot/npm_and_yarn/update_checker/version_resolver.rb]: nice cleanup. This one could perhaps be extracted to its own PR to keep the `pnpm` PR smaller?

File: npm_and_yarn/lib/dependabot/npm_and_yarn/update_checker/version_resolver.rb [hunk 9]
```diff
@@ -46,6 +46,15 @@ class VersionResolver
             YN0060:\s|\s.+\sprovides\s(?<required_dep>.+?)\s\((?<info_hash>\w+)\).+what\s(?<requiring_dep>.+?)\srequests
           /x
 
+        # Error message from pnpm add:
```
Comment (jeffwidman) [inline npm_and_yarn/lib/dependabot/npm_and_yarn/update_checker/version_resolver.rb:49]: Is this line stating that the error message from pnpm will look like the next two lines? Or that all three lines are what pnpm spits out?

File: npm_and_yarn/lib/dependabot/npm_and_yarn/update_checker/version_resolver.rb [hunk 10]
```diff
@@ -483,17 +498,21 @@ def lockfiles_for_path(lockfiles:, path:)
         end
 
         def run_checker(path:, version:)
-          # If there are both yarn lockfiles and npm lockfiles only run the
-          # yarn updater
           yarn_lockfiles = lockfiles_for_path(lockfiles: dependency_files_builder.yarn_locks, path: path)
           return run_yarn_checker(path: path, version: version, lockfile: yarn_lockfiles.first) if yarn_lockfiles.any?
 
+          pnpm_lockfiles = lockfiles_for_path(lockfiles: dependency_files_builder.pnpm_locks, path: path)
+          return run_pnpm_checker(path: path, version: version) if pnpm_lockfiles.any?
```
Comment (jeffwidman) [inline npm_and_yarn/lib/dependabot/npm_and_yarn/update_checker/version_resolver.rb:520]: Why does yarn win over pnpm if both are present? Why not check for presence of all three, and if more than one found, raise an error that we aren't sure what we should do?

There may be a good logical reason to still do what we do today, but if so, a code comment would be helpful to explain why.

File: npm_and_yarn/lib/dependabot/npm_and_yarn/file_fetcher.rb [hunk 11]
```diff
@@ -566,6 +566,12 @@ def ignore_package_lock?
         npmrc.content.match?(/^package-lock\s*=\s*false/)
       end
 
+      def ignore_pnpm_lock?
```
Comment (jeffwidman) [inline npm_and_yarn/lib/dependabot/npm_and_yarn/file_fetcher.rb:570]: The word `ignore` confused me for a sec, as I thought we somehow supported that in https://docs.github.com/en/code-security/dependabot/dependabot-version-updates/configuration-options-for-the-dependabot.yml-file#ignore

Looks like this is instead actually trying to respect this setting: https://pnpm.io/npmrc#lockfile

Maybe call it `skip_pnpm_lock?` to avoid confusion?

I'm also fine as-is...

File: npm_and_yarn/lib/dependabot/npm_and_yarn/update_checker/version_resolver.rb [hunk 12]
```diff
@@ -302,26 +304,34 @@ def updated_types_dependencies
         end
 
         def peer_dependency_errors
-          return @peer_dependency_errors if @peer_dependency_errors_checked
+          resolution_errors.reject { |error| error == "NOT_RESOLVABLE" }
```
Comment (jeffwidman) [inline npm_and_yarn/lib/dependabot/npm_and_yarn/update_checker/version_resolver.rb:307]: Where does this `"NOT_RESOLVABLE"` come from? pnpm itself, or something within Dependabot?

**Review by jeffwidman — COMMENTED** [review]
Took an initial review pass. It was very helpful that you'd kept a clean commit history so I could go through commit-by-commit to understand what the intent was.

Looks really solid so far, great job with the test cases covering a lot of potential error conditions.

A few comments, and some questions about stuff I didn't fully understand.

Also, I merged a bunch of your other PR's, so if you rebase it might slim this down a little?

File: npm_and_yarn/lib/dependabot/npm_and_yarn/file_fetcher.rb (hunk 2 above)
Comment (deivid-rodriguez) [reply npm_and_yarn/lib/dependabot/npm_and_yarn/file_fetcher.rb:444]: I think it is yeah.

Users may specify globs in their `pnpm-workspace.yml`, but it could happen that one of the folders that result from resolving the glob does not have a `package.json` file. That would be what this is rescuing. This is essentially copied from the logic of parsing the `workspaces` declaration in the `package.json` file which for `pnpm` lives in a separate file.

To be noted though, is that, we could probably use `Dir.glob` here directly (because we have the full repo cloned locally in this case). If we did that, there'd be no need for this rescue I think.

This was mentioned by @jakecoffman at https://github.com/dependabot/dependabot-core/pull/6932, and I suggested #6923 as a good moment to tackle it.

File: npm_and_yarn/lib/dependabot/npm_and_yarn/file_parser/lockfile_parser.rb (hunk 3 above)
Comment (deivid-rodriguez) [reply npm_and_yarn/lib/dependabot/npm_and_yarn/file_parser/lockfile_parser.rb:74]: Definitely.

As you mentioned, there are quite a few places were we establish a package manager priority, unfortunately.

In my opinion, it'd be probably be best to infer a package manager in the file fetcher, and error out there if there's a conflict. And then stop looking at different possibilities everywhere else and then just follow the chosen package manager logic through all the update assuming only the proper files for that package manager are present.

But that'd be probably a separate refactoring task?

File: npm_and_yarn/spec/dependabot/npm_and_yarn/file_parser/lockfile_parser_spec.rb (hunk 5 above)
Comment (deivid-rodriguez) [reply npm_and_yarn/spec/dependabot/npm_and_yarn/file_parser/lockfile_parser_spec.rb:113]: Yes, I think I meant

```suggestion
      context "locked to versions with peer disambiguation suffix" do
```

Will update!

File: npm_and_yarn/spec/dependabot/npm_and_yarn/file_parser/lockfile_parser_spec.rb (hunk 6 above)
Comment (deivid-rodriguez) [reply npm_and_yarn/spec/dependabot/npm_and_yarn/file_parser/lockfile_parser_spec.rb:128]: This is the error when failing to parse the lockfile (`pnpm-lock.yaml`). Your suggestion was to also raise it when failing to parse the `pnpm-workspace.yaml`, which makes sense to me and I will implement it!

File: npm_and_yarn/lib/dependabot/npm_and_yarn/update_checker/version_resolver.rb (hunk 8 above)
Comment (deivid-rodriguez) [reply npm_and_yarn/lib/dependabot/npm_and_yarn/update_checker/version_resolver.rb]: Sure.

File: npm_and_yarn/lib/dependabot/npm_and_yarn/update_checker/version_resolver.rb (hunk 10 above)
Comment (deivid-rodriguez) [reply npm_and_yarn/lib/dependabot/npm_and_yarn/update_checker/version_resolver.rb:520]: Yeah, same conversation as above. I think we should simplify this. I just tried to do something that follows the current pattern, but I think I also failed to make it consistent anyways.

I'm going to dedicate a small amount of time to try clean this up.

File: npm_and_yarn/lib/dependabot/npm_and_yarn/file_fetcher.rb (hunk 11 above)
Comment (deivid-rodriguez) [reply npm_and_yarn/lib/dependabot/npm_and_yarn/file_fetcher.rb:570]: I just copied it from the analogous `npm` setting, but I can rename both.

File: npm_and_yarn/lib/dependabot/npm_and_yarn/update_checker/version_resolver.rb (hunk 9 above)
Comment (deivid-rodriguez) [reply npm_and_yarn/lib/dependabot/npm_and_yarn/update_checker/version_resolver.rb:49]: The next two lines. It keeps the comment format of the constants for Yarn and NPM, but I agree it's not very clear. I think we could remove the `Error message from yarn add:`, `Error message from pnpm add:`, `Error message from npm install` headers? It also makes the comments easier to maintain because I actually stopped using `pnpm add` in favor of `pnpm update` and forgot to update this comment.

File: npm_and_yarn/lib/dependabot/npm_and_yarn/file_fetcher.rb [hunk 13]
```diff
@@ -172,6 +172,25 @@ def get_yarn_version_from_package_json(package_manager)
         version_match&.named_captures&.fetch("version", nil)
       end
 
+      def pnpm_version
+        return @pnpm_version if defined?(@pnpm_version)
+
+        if (package_manager = parsed_package_manager)
+          get_pnpm_version_from_package_json(package_manager)
+        elsif pnpm_lock
+          8
```
[the remark this answers is not in the document]
Comment (jeffwidman) [reply npm_and_yarn/lib/dependabot/npm_and_yarn/file_fetcher.rb:184]: Can you add a code comment saying that?

File: npm_and_yarn/lib/dependabot/npm_and_yarn/file_fetcher.rb [hunk 14]
```diff
@@ -523,6 +552,14 @@ def parsed_shrinkwrap
         {}
       end
 
+      def parsed_pnpm_workspace_yaml
+        return {} unless pnpm_workspace_yaml
+
+        YAML.safe_load(pnpm_workspace_yaml.content)
+      rescue Pysch::SyntaxError
+        {}
```
[the remark this answers is not in the document]
Comment (jeffwidman) [reply npm_and_yarn/lib/dependabot/npm_and_yarn/file_fetcher.rb:561]: Out of scope for this PR: your explanation for why to _not_ raise at this point for npm/yarn is great, and would be a very useful code comment in the relevant sections for npm/yarn...

File: npm_and_yarn/lib/dependabot/npm_and_yarn/file_fetcher.rb (hunk 2 above)
Comment (jeffwidman) [reply npm_and_yarn/lib/dependabot/npm_and_yarn/file_fetcher.rb:444]: If it's going away, then sounds fine... but if it's sticking around, explaining the expected context for why `nil` might be encountered is a useful code comment... that way if someone is debugging and hits nil, they can compare it to their scenario to see if it's an expected vs unexpected nil

File: npm_and_yarn/lib/dependabot/npm_and_yarn/file_parser/lockfile_parser.rb (hunk 3 above)
Comment (jeffwidman) [reply npm_and_yarn/lib/dependabot/npm_and_yarn/file_parser/lockfile_parser.rb:74]: Sounds reasonable on all counts.

File: npm_and_yarn/lib/dependabot/npm_and_yarn/file_fetcher.rb (hunk 14 above)
[the remark this answers is not in the document]
Comment (deivid-rodriguez) [reply npm_and_yarn/lib/dependabot/npm_and_yarn/file_fetcher.rb:561]: Agreed, will do!

File: npm_and_yarn/lib/dependabot/npm_and_yarn/update_checker/version_resolver.rb (hunk 9 above)
Comment (jeffwidman) [reply npm_and_yarn/lib/dependabot/npm_and_yarn/update_checker/version_resolver.rb:49]: I'm fine with whatever, titles can be helpful context but also understand dropping it as long as it's clear contextually where this error arises... I'm speaking with lots of scar tissue because over in the PHP checker we've got a bunch of error messages, but many of them are outdated with the switch to `composer 2`, only it's not clear which ones... if you choose to keep the title, do something to make it clear :
```suggestion
# Error message returned by `pnpm update` for `pnpm` version 8:
```

File: npm_and_yarn/lib/dependabot/npm_and_yarn/update_checker/version_resolver.rb (hunk 10 above)
Comment (jeffwidman) [reply npm_and_yarn/lib/dependabot/npm_and_yarn/update_checker/version_resolver.rb:520]: I'm cool with also just filing an issue, cross-linking it to our upcoming "ecosystem clarity" effort and moving on for now.

The other way to solve this long term is if we get the user to explicitly tell us which package manager they want us to use... `pnpm`, `yarn`, `npm`, rather than us guessing... I prefer that, although we'll need to figure out how that works in the context of security updates where a repo may not have a `dependabot.yaml` config file.

But if there's easy enough cleanup/tidying where it clarifies the code without getting all the way into redoing our config files, then yeah, that makes sense too... I trust your judgement.

File: npm_and_yarn/helpers/lib/pnpm/lockfile-parser.js [hunk 15]
```diff
@@ -0,0 +1,73 @@
+/* PNPM-LOCK.YAML PARSER
+ *
+ * Inputs:
+ *  - directory containing a pnpm-lock.yaml file
+ *
+ * Outputs:
+ *  - JSON formatted information of dependencies (name, version, type)
```
[the remark this answers is not in the document]
Comment (deivid-rodriguez) [reply npm_and_yarn/helpers/lib/pnpm/lockfile-parser.js]: Added to 876e9d74854f2d09d2d46440309aa943f609064e!

File: npm_and_yarn/lib/dependabot/npm_and_yarn/file_fetcher.rb (hunk 14 above)
[the remark this answers is not in the document]
Comment (deivid-rodriguez) [reply npm_and_yarn/lib/dependabot/npm_and_yarn/file_fetcher.rb]: I changed 876e9d74854f2d09d2d46440309aa943f609064e to raise a `DependencyFileNotParseable` error. I was going to create a separate PR to comment on why we swallow errors when parsing lockfiles here, but I'm not sure it's the right thing to do vs raising directly, so I decided to pass and move on to other things.

File: npm_and_yarn/lib/dependabot/npm_and_yarn/file_parser/lockfile_parser.rb (hunk 3 above)
Comment (deivid-rodriguez) [reply npm_and_yarn/lib/dependabot/npm_and_yarn/file_parser/lockfile_parser.rb:72]: I did change the priority so that it's at least consistent for now in the different places.

File: npm_and_yarn/lib/dependabot/npm_and_yarn/update_checker/version_resolver.rb (hunk 10 above)
Comment (deivid-rodriguez) [reply npm_and_yarn/lib/dependabot/npm_and_yarn/update_checker/version_resolver.rb:520]: I opened a ticket! https://github.com/dependabot/dependabot-core/issues/7184

File: npm_and_yarn/lib/dependabot/npm_and_yarn/update_checker/version_resolver.rb (hunk 8 above)
Comment (deivid-rodriguez) [reply npm_and_yarn/lib/dependabot/npm_and_yarn/update_checker/version_resolver.rb]: Done at #7178!

File: npm_and_yarn/lib/dependabot/npm_and_yarn/file_fetcher.rb (hunk 2 above)
Comment (deivid-rodriguez) [reply npm_and_yarn/lib/dependabot/npm_and_yarn/file_fetcher.rb:445]: I added a comment a refactored this a bit at #7183.

File: npm_and_yarn/lib/dependabot/npm_and_yarn/file_fetcher.rb (hunk 13 above)
[the remark this answers is not in the document]
Comment (jeffwidman) [reply npm_and_yarn/lib/dependabot/npm_and_yarn/file_fetcher.rb]: Nice! Best code comments are well-named method/var names. 😁

File: npm_and_yarn/lib/dependabot/npm_and_yarn/update_checker/version_resolver.rb (hunk 12 above)
Comment (deivid-rodriguez) [reply npm_and_yarn/lib/dependabot/npm_and_yarn/update_checker/version_resolver.rb]: I removed this commit. It was an improvement for an edge case (trying to update to a version with yanked deps), but it's not specific to PNPM, so I'll propose it separately if needed.

File: npm_and_yarn/lib/dependabot/npm_and_yarn/update_checker/version_resolver.rb (hunk 9 above)
Comment (deivid-rodriguez) [reply npm_and_yarn/lib/dependabot/npm_and_yarn/update_checker/version_resolver.rb]: I ended up updating the message and aligning the other headers at #7195.

File: npm_and_yarn/lib/dependabot/npm_and_yarn/file_fetcher.rb (hunk 11 above)
Comment (deivid-rodriguez) [reply npm_and_yarn/lib/dependabot/npm_and_yarn/file_fetcher.rb]: Renamed and aligned analogous method for NPM at #7197.

**Comment by deivid-rodriguez** [comment]
This is ready for another review round @jeffwidman!

**Review by AntoineKM — COMMENTED** [review]
Impressive work, well done, love it !

File: npm_and_yarn/lib/dependabot/npm_and_yarn/file_fetcher.rb [hunk 16]
```diff
@@ -496,6 +509,12 @@ def skip_package_lock?
         npmrc.content.match?(/^package-lock\s*=\s*false/)
       end
 
+      def skip_pnpm_lock?
+        return false unless npmrc
+
+        npmrc.content.match?(/^lockfile\s*=\s*false/)
```
Comment (jeffwidman) [inline npm_and_yarn/lib/dependabot/npm_and_yarn/file_fetcher.rb:556]: Could this `lockfile` key have whitespace in front of it and still work?

File: npm_and_yarn/lib/dependabot/npm_and_yarn/file_fetcher.rb [hunk 17]
```diff
@@ -169,6 +170,18 @@ def guess_yarn_version
         Helpers.yarn_version_numeric(yarn_lock)
       end
 
+      def pnpm_version
+        return @pnpm_version if defined?(@pnpm_version)
+
+        @pnpm_version = package_manager.locked_version("pnpm") || guess_pnpm_version
+      end
+
+      def guess_pnpm_version
```
Comment (jeffwidman) [inline npm_and_yarn/lib/dependabot/npm_and_yarn/file_fetcher.rb:182]: I really like the refactor/rename you did with this method, it's so much more clear the intent. 👍

**Review by jeffwidman — APPROVED** [review]
Looks great! 🎉 

Had a few trivial comments, they do not block shipping, so can either be addressed now or in a follow-on PR since I know we're trying to get this live so beta users can start hammering on it.

File: npm_and_yarn/lib/dependabot/npm_and_yarn/file_fetcher.rb (hunk 16 above)
Comment (deivid-rodriguez) [reply npm_and_yarn/lib/dependabot/npm_and_yarn/file_fetcher.rb:556]: I don't know, I followed the pattern of `skip_package_lock?`. This is an INI file, and I'm not sure if that's valid.

That said, this feature is questionable, since it'd be only useful if someone is committing a lockfile and configuring it to not be used. Maybe overkill to add a special case for this ourselves, so we could alternatively remove `skip_package_lock?` and not add this.

File: npm_and_yarn/lib/dependabot/npm_and_yarn/file_parser/lockfile_parser.rb [hunk 18]
```diff
@@ -78,6 +81,12 @@ def package_locks
             select { |f| f.name.end_with?("package-lock.json") }
         end
 
+        def pnpm_locks
+          @pnpm_locks ||=
+            dependency_files.
+            select { |f| f.name.end_with?("pnpm-lock.yaml") }
```
[the remark this answers is not in the document]
Comment (deivid-rodriguez) [reply npm_and_yarn/lib/dependabot/npm_and_yarn/file_parser/lockfile_parser.rb:87]: Yep, it's exactly that `== <name>` is used for root files, `.end_with?(<name>)` for any file with a particular name.

We could probably improve this but it's a pattern used all over, so I just followed the existing pattern.

File: npm_and_yarn/lib/dependabot/npm_and_yarn/update_checker/version_resolver.rb [hunk 19]
```diff
@@ -46,6 +46,15 @@ class VersionResolver
             YN0060:\s|\s.+\sprovides\s(?<required_dep>.+?)\s\((?<info_hash>\w+)\).+what\s(?<requiring_dep>.+?)\srequests
           /x
 
+        # Error message returned by `pnpm update`:
```
[the remark this answers is not in the document]
Comment (deivid-rodriguez) [reply npm_and_yarn/lib/dependabot/npm_and_yarn/update_checker/version_resolver.rb:49]: I have not observed any variations across versions, so I went with optimistically assuming it's stable (at least so far).

File: npm_and_yarn/lib/dependabot/npm_and_yarn/file_updater/pnpm_lockfile_updater.rb [hunk 20]
```diff
@@ -0,0 +1,172 @@
...
+          handle_pnpm_lock_updater_error(e, pnpm_lock)
+        end
+
+        private
+
+        attr_reader :dependencies, :dependency_files, :repo_contents_path, :credentials
+
+        IRRESOLVABLE_PACKAGE = "ERR_PNPM_NO_MATCHING_VERSION"
+        INVALID_REQUIREMENT = "ERR_PNPM_SPEC_NOT_SUPPORTED_BY_ANY_RESOLVER"
+        MISSING_PACKAGE = /(?<package_req>.*?) is not in the npm registry, or you have no permission to fetch it/
```
[the remark this answers is not in the document]
Comment (deivid-rodriguez) [reply npm_and_yarn/lib/dependabot/npm_and_yarn/file_updater/pnpm_lockfile_updater.rb:37]: That'd be any supported version I think, since I haven't seen any differences between versions. Also, I believe in particular the `ERR_PNPM` codes are there to allow changing the error wording without changing the error code, so those should stay stable. Unfortunately, not all errors have codes like that.

File: npm_and_yarn/lib/dependabot/npm_and_yarn/file_fetcher.rb (hunk 16 above)
Comment (jeffwidman) [reply npm_and_yarn/lib/dependabot/npm_and_yarn/file_fetcher.rb:556]: I'm also cool with shipping as-is, and we can always later on add extra handling for whitespace if that becomes a problem. My guess is it never will.

File: npm_and_yarn/lib/dependabot/npm_and_yarn/file_updater/pnpm_lockfile_updater.rb (hunk 20 above)
[the remark this answers is not in the document]
Comment (jeffwidman) [reply npm_and_yarn/lib/dependabot/npm_and_yarn/file_updater/pnpm_lockfile_updater.rb:37]: Ah, that's brilliant of the `pnpm` folks to ship well-known error codes!

File: npm_and_yarn/lib/dependabot/npm_and_yarn/update_checker/version_resolver.rb (hunk 19 above)
[the remark this answers is not in the document]
Comment (jeffwidman) [reply npm_and_yarn/lib/dependabot/npm_and_yarn/update_checker/version_resolver.rb:49]: Sounds fine, I'm not so confident it'll always be stable, but there's always the `git blame` time machine for figuring this out...

**Comment by deivid-rodriguez** [comment]
Alright, let's try this!

**Comment by AntoineKM** [comment]
Hi, I just wanted to know when it will be published on @dependabot ?

**Comment by BrianLovelace128** [comment]
Also interested in the answer to AntoineKM's question.

**Comment by deivid-rodriguez** [comment]
Hello, sorry for the delay. The enablement of this feature has been delayed slightly, but will be made generally available soon. I will post a more specific update once I know all the details!

**Comment by Stanzilla** [comment]
@deivid-rodriguez looks like it is out now? https://github.com/dependabot/dependabot-core/releases/tag/v0.218.0

**Comment by PeterMK85** [comment]
>it doesn't work yet.

Same here.

```bash
The property '#/updates/0/package-ecosystem' value "pnpm" did not match one of the following values: npm, bundler, composer, maven, mix, cargo, gradle, nuget, gomod, docker, elm, gitsubmodule, github-actions, pip, terraform, pub
```

`.github/dependabot.yml`

```yaml
version: 2
enable-beta-ecosystems: true
updates:
    - package-ecosystem: 'pnpm'
```

**Comment by mhogeveen** [comment]
@PeterMK85 For the beta testing the `package-ecosystem` key needed to be set to `npm`, not `pnpm`. Maybe then it will work for you?

**Comment by PeterMK85** [comment]
@mhogeveen 

Okay so adjusting `pnpm` to `npm` started to work 

_but_ 

There is a major issue:
Only `package.json` is being updated but `pnpm-lock.yaml` does not in a monorepo pnpm setup.
So it will fail on the next installation:
`ERR_PNPM_OUTDATED_LOCKFILE  Cannot install with "frozen-lockfile" because pnpm-lock.yaml is not up to date with package.json`

**Comment by Rawnly** [comment]
Same problem here, monorepo with pnpm, lock file is ignored

**Comment by mhogeveen** [comment]
@PeterMK85

All beta testers I checked do not have the problem you are having. The key differences I've noticed are the use of the `packageManager` key in the `package.json` and the fact that they were/are beta testers.

You can check out this issue (https://github.com/dependabot/dependabot-core/issues/1736) to find the beta testers repo's for examples of dependabot working on pnpm monorepo's.

Might be worth adding the `packageManager` key to see if that works. Otherwise I'm out of ideas 🤷‍♂️

**Comment by Rawnly** [comment]
Hey @mhogeveen i have the same problem as Peter but I'm currently using the package manager key in my package json. It's currently pnpm 8.5.1 and I'm in the beta. 

All the PRs created by dependabot are failing checks and can't be merged due to the lockfile not being up to date. 

Let me know if you need any other information. 

PS: we're currently on a monorepo managed by turbo.

**Comment by PeterMK85** [comment]
@mhogeveen 

Is it still required to send an email with the repo to @carogalvin ? 
I'm not part of the Beta I thought it was publicly rolled out.

**Comment by mhogeveen** [comment]
Just a quick disclaimer: I'm not affiliated to the project! Just a beta tester trying to help out. If none of the options I've mentioned above seem to work, it's probably best to wait until there is an official message from staff or a contributor.

**Comment by Rawnly** [comment]
> Is it still required to send an email with the repo to @carogalvin ?

this is what we've done so far. Try to reach her out and see if it's still needed

**Comment by Nishnha** [comment]
This feature is still in beta. Unfortunately, we are not accepting more beta testers at the moment.

We will post a larger announcement when pnpm support is generally available, so please be patient and thank you for your support!

**Comment by PeterMK85** [comment]
@Nishnha 

Thanks for the update, by and chance can you share the planned public roll out date(range)?

If it's a matter of weeks / months, might affect the technical approach to resolve the issue.

Huge thanks again for the effort

**Comment by AntoineKM** [comment]
@PeterMK85 if you're looking for an alternative in the meantime you might try @renovate

**Comment by carogalvin** [comment]
Hey everyone! Sorry for the current frustration with the rollout, we're in a company-wide change freeze so we were unfortunately halted on enabling this when we wanted to originally. We are ready to go to GA immediately when it lifts, and we'll let you know on the [issue](https://github.com/dependabot/dependabot-core/issues/1736) when that happens. I can't really give you a date because the target date keeps shifting and I don't want to accidentally lie to you 😅 As a rough estimate, I expect more than a week but less than a month.

We are also investigating the problem with lockfiles!
